### PR TITLE
Fix log-Euclidean Log and Exp maps

### PR DESCRIPTION
**Pull request by gabelstein**
Fix of the logeuclid exp and log maps. 

Needed to add a first_divided_difference function. Not sure where to put it, lives in .utils.base for now.

### CODE REVIEW

File: pyriemann/utils/tangentspace.py [hunk 1]
```diff
@@ -99,11 +99,8 @@ def exp_map_logeuclid(X, Cref):
 
     The projection of a matrix :math:`\mathbf{X}` from tangent space
     to SPD/HPD manifold with log-Euclidean exponential map
-    according to a reference SPD/HPD matrix :math:`\mathbf{C}_\text{ref}` is:
-
-    .. math::
-        \mathbf{X}_\text{original} =
-        \exp(\mathbf{X} + \log(\mathbf{C}_\text{ref}))
+    according to a reference SPD/HPD matrix :math:`\mathbf{C}_\text{ref}` as
+    described in [1]_.
```
Comment (qbarthelemy) [inline pyriemann/utils/tangentspace.py:103]: We should cite Eq.(3.4) of Arsigny2007, _Geometric means in a novel vector space structure on symmetric positive-definite matrices_
```suggestion
    described in Eq.(3.4) of [1]_:

    .. math::
        \mathbf{X}_\text{original} = \exp \left( \log(\mathbf{C}_\text{ref})
        + D_{\mathbf{C}_\text{ref}} \log . \mathbf{X} \right)

    where :math:`[D_{\mathbf{C}_\text{ref}} \log . \mathbf{X}` indicates the
    directional derivative of the matrix logarithm at
    :math:`\mathbf{C}_\text{ref}` along :math:`X`.
```

File: pyriemann/utils/tangentspace.py [hunk 2]
```diff
@@ -247,11 +253,8 @@ def log_map_logeuclid(X, Cref):
     r"""Project matrices in tangent space by log-Euclidean logarithmic map.
 
     The projection of a matrix :math:`\mathbf{X}` from SPD/HPD manifold
-    to tangent space by log-Euclidean logarithmic map
-    according to a SPD/HPD reference matrix :math:`\mathbf{C}_\text{ref}` is:
-
-    .. math::
-        \mathbf{X}_\text{new} = \log(\mathbf{X}) - \log(\mathbf{C}_\text{ref})
+    to tangent space by log-Euclidean logarithmic map according to a SPD/HPD
+    reference matrix :math:`\mathbf{C}_\text{ref}` as described in [1]_.
```
Comment (qbarthelemy) [inline pyriemann/utils/tangentspace.py:257]: Similarly, we should cite Eq.(3.4) of Arsigny2007 and give equation.

**Review by qbarthelemy — CHANGES_REQUESTED** [review]
Thanks Gabriel for this crucial correction!

File: pyriemann/utils/tangentspace.py (hunk 1 above)
Comment (gabelstein) [reply pyriemann/utils/tangentspace.py:103]: Yes, that would be the proper original citation, but the paper does not show how to compute the directional derivative. Not sure if that helps.

File: pyriemann/utils/tangentspace.py [hunk 3]
```diff
@@ -120,8 +117,17 @@ def exp_map_logeuclid(X, Cref):
...
+    ----------
+    .. [1] `A New Canonical Log-Euclidean Kernel for Symmetric Positive
+        Definite Matrices for EEG Analysis
+        <https://ieeexplore.ieee.org/document/10735221>`_
+        G. Wagner vom Berg, V. Röhr, D. Platt, B. Blankertz. IEEE TBME, 2024.
     """
-    return expm(X + logm(Cref))
+    d, V = np.linalg.eigh(Cref)
+    logfdd = _first_divided_difference(d, np.log, lambda x: 1 / x)
+    return expm(logm(Cref) + V @ (logfdd * (V.conj().T @ X @ V)) @ V.conj().T)
```
Comment (qbarthelemy) [inline pyriemann/utils/tangentspace.py:130]: I think it's interesting to add a function to calculate the directional derivative of the matrix logarithm:
```
def ddlogm(X, Cref):
    d, V = np.linalg.eigh(Cref)
    logfdd = _first_divided_difference(d, np.log, lambda x: 1 / x)
    return V @ (logfdd * (V.conj().T @ X @ V)) @ V.conj().T
```
Same remark for the directional derivative of the matrix exponential.

**Review by qbarthelemy — CHANGES_REQUESTED** [review]

File: pyriemann/utils/base.py [hunk 4]
```diff
@@ -232,3 +232,143 @@ def nearest_sym_pos_def(X, reg=1e-6):
...
+    :math:`\mathbf{X}` is defined as Eq. (V.13) in [1]_:
+
+    .. math::
+        \text{ddexpm}(\mathbf{X}, \mathbf{C}_{\text{ref}}) =
+        \mathbf{V} \left( \text{fddexpm}(\mathbf{\Lambda}) \odot
+        \mathbf{V}^H \mathbf{X} \mathbf{V} \right) \mathbf{V}^H
+
+    where :math:`\mathbf{\Lambda}` is the diagonal matrix of eigenvalues of
+    and :math:`\mathbf{V}` the eigenvectors of :math:`\mathbf{C}_{\text{ref}}`,
+    and :math:`\text{fddexpm}` the first divided difference of the exponential function.
```
Comment (qbarthelemy) [inline pyriemann/utils/base.py:348]:
```suggestion
    and :math:`\text{fddexpm}` the first divided difference of the exponential
    function.
```

File: pyriemann/utils/base.py [hunk 5]
```diff
@@ -232,3 +232,143 @@ def nearest_sym_pos_def(X, reg=1e-6):
...
+    :math:`\mathbf{X}` is defined as Eq. (V.13) in [1]_:
+
+    .. math::
+        \text{ddlogm}(\mathbf{X}, \mathbf{C}_{\text{ref}}) =
+        \mathbf{V} \left( \text{fddlogm}(\mathbf{\Lambda}) \odot
+        \mathbf{V}^H \mathbf{X} \mathbf{V} \right) \mathbf{V}^H
+
+    where :math:`\mathbf{\Lambda}` is the diagonal matrix of eigenvalues
+    and :math:`\mathbf{V}` the eigenvectors of :math:`\mathbf{C}_{\text{ref}}`,
+    and :math:`\text{fddlogm}` the first divided difference of the logarithm function.
```
Comment (qbarthelemy) [inline pyriemann/utils/base.py:305]:
```suggestion
    and :math:`\text{fddlogm}` the first divided difference of the logarithm
    function.
```

File: pyriemann/utils/tangentspace.py [hunk 6]
```diff
@@ -120,8 +126,22 @@ def exp_map_logeuclid(X, Cref):
...
 
 
 def exp_map_riemann(X, Cref, Cm12=False):
```
Comment (qbarthelemy) [inline pyriemann/utils/tangentspace.py:147]: What reference could we add to this function?

File: pyriemann/utils/base.py [hunk 7]
```diff
@@ -232,3 +232,148 @@ def nearest_sym_pos_def(X, reg=1e-6):
...
+    dif = np.repeat(d[None, :], len(d), axis=0)
+
+    close_ = np.isclose(dif, dif.T, atol=atol, rtol=rtol)
+    dif[close_] = fctder(dif[close_])
+    dif[~close_] = (fct(dif[~close_]) - fct(dif.T[~close_])) / \
+                   (dif[~close_] - dif.T[~close_])
+    return dif
+
+
+def ddlogm(X, Cref):
```
Comment (qbarthelemy) [inline pyriemann/utils/base.py:294]: To respect alphabetic order, can you declare `ddexpm` before `ddlogm`?

**Review by qbarthelemy — APPROVED** [review]
Thx @gabelstein for this correction!

We can merge when #335 is ok.

**Comment by qbarthelemy** [comment]
Thx @gabelstein !

**Pull request merged**